### PR TITLE
Add a joule thief example

### PR DESCRIPTION
**Pull request by kasbah**
This is a start of an attempt to add a joule thief example.

Some notes:

- Was very confused about all the `*` importing, no idea where anything came from
- Was confused about footprints as I wanted to assign a footprint to my "device" before I looked at the comments in footprints.py, i think `Package` should be called `Footprint` and the combination should be `Part` or `Component` or something like that.
- Still confused about whether I should be using devices or footprints in my circuit
- I really like the for loop to generate all the SOT23 transistor footprints!

Will have a go at making some "packages" for the choke and the battery connector when I find time. Any pointers or advice appreciated.

### CODE REVIEW

**Comment by dvc94ch** [comment]
Looking good. Hopefully writting the packages shouldnt be too hard, there are a couple of examples in the library. Proper routing with monosat is still a couple of days of development effort, and requires some experimentation. Input and output formats need to change and the router copied from monosat examples needs to be rewritten. Maybe this weekend.

**Comment by kasbah** [comment]
Ok, so I managed to "draw" the footprint for an TDK  ACT45B

image

```python
Package('TDK:ACT45B', RectCrtyd(5.9, 3.4), DualPads(4, 2.5, radius=2.275),
        package_size=(5.9, 3.4), pad_size=(0.9, 1.35))
```

I can't say the process was very intuitive, I am still not sure how to do it, I just fiddled with it till it looked right. The term "radius" seems especially confusing. Would you have defined this in a different way?

I ran the placing and it came up with this: 

image

**Comment by kasbah** [comment]
Not sure, why I am not seeing any connections for pads 1 and 4 of the choke.

**Comment by kasbah** [comment]
Ah, figured out that I can't just connect pins to pins.

**Comment by kasbah** [comment]
Ok, this is done as an example. Not sure you need another simple example but it helped me understand everything better. I could put TDK ACT45B into the libraries but am not sure where.

File: examples/joule_thief/joule_thief.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+
+for a, b, c in ('BCE', 'BEC', 'CBE', 'CEB', 'ECB', 'EBC'):
+    Footprint('SOT23' + a + b + c, 'Q', 'SOT23',
+              Map(1, a),
+              Map(2, b),
+              Map(3, c))
+
+
+Device('Transformer_1P_1S', pins=(Pin('1'), Pin('2'), Pin('3'), Pin('4')))
```
Comment (dvc94ch) [inline examples/joule_thief/joule_thief.py]: I'd prefer the pins to be named like this:
Pin('L1.1', 'L1'), Pin('L1.2', 'L1'), Pin('L2.1', 'L2'), Pin('L2.2', 'L2')

'L1' and 'L2' are functions and mean that pin 1 and 2 are interchangeable

**Comment by dvc94ch** [comment]
That's a bug

**Comment by dvc94ch** [comment]
How did you find the process in general? Apart from the naming of things? Was it terribly annoying?

**Comment by kasbah** [comment]
At this stage, and for a circuit like this, I'd definitely prefer to just draw a schematic. I spotted a few glimpses of the potential for all of this though: 
- The for loop to associate footprints
- The succinctness of the `Package` expressions

Having said that, I haven't understood the terminology and approach for actually defining a package though I somehow managed to do it.

I really struggled with the name-spacing and mutation of everything. Something where there isn't this implicit "circuit design" floating about that is being changed all the time would suit me much better.

**Comment by kasbah** [comment]
To add to that: I think it's _super easy_ to make mistakes and not get reasonable error messages in response. Its early stages, so it's understandable, but I think it's worth considering for improvements or re-design of the HDL and the general flow.

**Comment by dvc94ch** [comment]
Those are my experiences too. Thank you for contributing the example!

**Pull request merged**